cscs-checks/microbenchmarks/shmem/src/shmem.cu

```diff
@@ -0,0 +1,107 @@
+
+// Shared memory bandwidth benchmark
+// contributed by Sebastian Keller
+// 
+// Relevant nvprof metrics:
+// nvprof -m shared_load_throughput,shared_store_throughput
+
+#include <iostream>
+
+#include <malloc.h>
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#define NTHREADS 256
+#define NITER    4096
+// must be even
+#define SHARED_SEGMENTS 4
+
+static void HandleError( cudaError_t err,
+                         const char *file,
+                         int line ) {
+    if (err != cudaSuccess) {
+        printf( "%s in %s at line %d\n", cudaGetErrorString( err ),
+                file, line );
+        exit( EXIT_FAILURE );
+    }
+}
+#define HANDLE_ERROR( err ) (HandleError( err, __FILE__, __LINE__ ))
+
+template <class T>
+__device__ void swap(T* a, T* b)
+{
+    T tmp;
+    tmp = *a;
+    *a = *b;
+    // +1 isn't needed to prevent code elimination by the
+    // compiler, but is added in case it gets smarter in
+    // a future version
+    *b = tmp + T(1);
+}
+
+template <class T>
+__global__ void test_shmem(T* glob_mem)
+{
+    __shared__ T smem[NTHREADS*SHARED_SEGMENTS];
+
+    int tid = threadIdx.x;
+
+    smem[tid] = T(1);
+    for (int i = 0; i < NITER; ++i)
+    {
+        // even shared segments
+        for (int j = 0; j < SHARED_SEGMENTS-1; j+=2)
+            swap(smem + tid + j*blockDim.x, smem + tid + (j+1)*blockDim.x);
+
+        // uneven shared segments
+        for (int j = 1; j < SHARED_SEGMENTS-2; j+=2)
+            swap(smem + tid + j*blockDim.x, smem + tid + (j+1)*blockDim.x);
+    }
+
+    glob_mem[blockIdx.x * blockDim.x + tid] = smem[tid];
+}
+
+template <class T>
+double test_bw(long size)
+{
+    T* buffer = (T*)malloc(size);
+    T* dev_buffer; 
+    HANDLE_ERROR( cudaMalloc((void**)&dev_buffer, size) );
+    int nblocks = size / (NTHREADS * sizeof(T));
```

Comment:
Because a single block doesn't do the work, neither theoretically nor experimentally.
Are you claiming that the device can be saturated with a single thread block of 256 threads?

Reply:
You are right, you cannot saturate it, because the SM wouldn't be full. What is your occupancy now?

Reply:
97.25%